### PR TITLE
GeoTransolver GALE: geometry-latent cross-attention in TransolverBlock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The core OOD failure on p_tan is that NACA6416 (test) has different camber, chord and thickness distributions from NACA0012 (training). The backbone sees dsdf features that encode proximity to geometry, but no explicit geometric "shape identity" signal is injected into the global slice tokens. GeoTransolver (arXiv 2412.14171) showed that injecting a geometry latent into the Transolver attention blocks via cross-attention significantly improves OOD generalization on airfoil shape families.

**Mechanism:** Pool surface node features → fixed-size geometry latent vector (per foil) → cross-attend slice tokens against geometry latent at each TransolverBlock. The slice tokens now carry explicit shape conditioning, so the physics-attention routing adapts to the geometry being processed rather than relying solely on node-local dsdf features.

The TE coordinate frame (PR #2207, merged, -5.4% p_in) confirmed that adding explicit geometric reference signals helps OOD. This is the next level: instead of fixed geometric offsets, inject a **learned** geometric representation into the global routing mechanism.

**Source:** arXiv 2412.14171 (GeoTransolver), human team requested via issue #1926.

**Target:** p_tan primarily (NACA6416 OOD), secondarily p_oodc.

### Step 1: Add config flags

In the `Config` dataclass:
```python
geotransolver_gale: bool = False  # Geometry-latent cross-attention conditioning
gale_latent_dim: int = 32         # Geometry latent dimension per foil
```

### Step 2: Add the GeometryEncoder module

Add this new module near the other model classes:
```python
class GeometryEncoder(nn.Module):
    """Pool surface node features into a fixed-size geometry latent."""
    def __init__(self, in_dim, latent_dim):
        super().__init__()
        self.proj = nn.Sequential(
            nn.Linear(in_dim, latent_dim * 2),
            nn.GELU(),
            nn.Linear(latent_dim * 2, latent_dim),
        )
    def forward(self, surface_x, surface_mask):
        # surface_x: [B, N, D], surface_mask: [B, N] bool
        masked = surface_x * surface_mask.unsqueeze(-1).float()
        count = surface_mask.float().sum(1, keepdim=True).clamp(min=1)
        pooled = masked.sum(1) / count  # [B, D]
        return self.proj(pooled)  # [B, latent_dim]
```

Separately encode fore-foil and aft-foil surface nodes, concatenate → `geom_latent` of dim `2 * latent_dim`.

### Step 3: Add cross-attention to TransolverBlock

In `TransolverBlock.__init__`, add parameter `geotransolver_gale: bool = False` and `gale_latent_dim: int = 32`:
```python
self.geotransolver_gale = geotransolver_gale
if geotransolver_gale:
    gale_total_dim = gale_latent_dim * 2  # fore + aft foil latents concatenated
    self.gale_q_proj = nn.Linear(n_hidden, n_hidden)
    self.gale_k_proj = nn.Linear(gale_total_dim, n_hidden)
    self.gale_v_proj = nn.Linear(gale_total_dim, n_hidden)
    self.gale_out_proj = nn.Linear(n_hidden, n_hidden)
    # CRITICAL: Zero-init output projection so cross-attention starts as no-op
    nn.init.zeros_(self.gale_out_proj.weight)
    nn.init.zeros_(self.gale_out_proj.bias)
```

In `TransolverBlock.forward`, add a `geom_latent` parameter (default None). After the slice-attention (physics attention) step but BEFORE the FFN, apply geometry cross-attention:
```python
# After physics attention, before FFN:
if self.geotransolver_gale and geom_latent is not None:
    # geom_latent: [B, 2*latent_dim]
    # Apply cross-attention on the per-node hidden states fx directly:
    gale_q = self.gale_q_proj(fx)              # [B, N, D]
    gale_kv = geom_latent.unsqueeze(1)         # [B, 1, 2*latent_dim]
    gale_k = self.gale_k_proj(gale_kv)         # [B, 1, D]
    gale_v = self.gale_v_proj(gale_kv)         # [B, 1, D]
    cross_out = F.scaled_dot_product_attention(gale_q, gale_k, gale_v)  # [B, N, D]
    fx = fx + self.gale_out_proj(cross_out)    # additive, zero-init → no-op at start
```

**Important:** Cross-attention is between ALL node hidden states (queries) and a SINGLE geometry latent token (key/value). Computationally cheap — N×1 attention collapses to a linear projection scaled by attention weight.

**Also important:** The `adaln_all` code path (if present) also needs the same cross-attention insertion point. Check if there's a separate branch for `self.adaln_all` in the forward method and apply GALE there too.

### Step 4: Build geometry latent in the Transolver model

In `Transolver.__init__`, if `cfg.geotransolver_gale`:
```python
if cfg.geotransolver_gale:
    self.fore_geom_enc = GeometryEncoder(in_channels, cfg.gale_latent_dim)
    self.aft_geom_enc = GeometryEncoder(in_channels, cfg.gale_latent_dim)
```

In `Transolver.forward`, before the TransolverBlock loop, compute the geometry latent:
```python
if self.cfg.geotransolver_gale:
    # Use existing zone classification:
    # zone 1 = fore-foil surface, zone 2 = aft-foil surface
    fore_surf_mask = (zones == 1)  # [B, N]
    aft_surf_mask = (zones == 2)   # [B, N]
    
    fore_latent = self.fore_geom_enc(x, fore_surf_mask)  # [B, latent_dim]
    aft_latent = self.aft_geom_enc(x, aft_surf_mask)     # [B, latent_dim]
    geom_latent = torch.cat([fore_latent, aft_latent], dim=-1)  # [B, 2*latent_dim]
else:
    geom_latent = None
```

**For single-foil samples:** `aft_surf_mask` is all-False → aft encoder gets zero-pooled input → learns a "no aft foil" latent. This is correct — no special handling needed.

Pass `geom_latent` to each TransolverBlock:
```python
for block in self.blocks:
    fx = block(fx, ..., geom_latent=geom_latent)
```

### Step 5: Verify torch.compile compatibility

All ops (Linear, GELU, scaled_dot_product_attention, cat, masked sum) are torch.compile-safe. No dynamic shapes — geometry latent is always [B, 2*latent_dim]. Should compile cleanly.

### Step 6: Diagnostic logging

After training, log `||gale_out_proj.weight||_F` per block to W&B — reveals whether geometry conditioning learned to contribute or stayed near zero.

## Baseline

**Current single-model baseline (PR #2207, +TE Coordinate Frame, 2-seed avg):**

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **12.490** | < 12.49 |
| **p_oodc** | **7.618** | < 7.62 |
| **p_tan** | **28.521** | < 28.52 |
| **p_re** | **6.411** | < 6.41 |

W&B runs: obn1wfja (seed 42), 52irfwwg (seed 73).

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/baseline-te-coord" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

## Experiment

**Run 1 — single seed validation:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/geotransolver-gale-s42" \
  --wandb_group "geotransolver-gale" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --geotransolver_gale --gale_latent_dim 32 \
  --seed 42
```

**If seed 42 shows improvement, run seed 73:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/geotransolver-gale-s73" \
  --wandb_group "geotransolver-gale" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --geotransolver_gale --gale_latent_dim 32 \
  --seed 73
```

## Implementation Notes

- **Zero-init is critical.** The `gale_out_proj` must be zero-initialized so the model starts exactly at baseline behavior. Without this, the random cross-attention output will break convergence.
- **Surface masks:** Use the existing zone classification (zone 1 = fore surface, zone 2 = aft surface). Check how `surface_mask` is constructed in the existing code — the SRF heads already separate fore vs aft foil nodes.
- **Single-foil samples:** The aft geometry encoder receives all-zero pooled input. This naturally produces a distinct "no aft foil" latent. No special handling needed.
- **VRAM:** Cross-attention adds ~75K parameters across 3 blocks. The N×1 attention pattern adds negligible compute. VRAM impact should be <1GB.
- **torch.compile:** All ops are compile-safe. No dynamic control flow, no variable-length sequences.
- **Diagnostic:** After training, report `||gale_out_proj.weight||_F` per block — reveals whether geometry conditioning learned to contribute.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Seed 42 (126 epochs, 180.9 min)

| Metric | Baseline (2-seed avg) | GALE s42 | Delta |
|--------|----------------------|----------|-------|
| **p_in** | 12.490 | **15.442** | **+23.6% ↑ (worse)** |
| **p_tan** | 28.521 | **29.2** | **+2.4% ↑ (worse)** |
| **p_oodc** | 7.618 | **8.9** | **+16.8% ↑ (worse)** |
| **p_re** | 6.411 | **7.6** | **+18.5% ↑ (worse)** |

**Full metrics:**
- val_in_dist: loss=0.2991, mae_surf_Ux=3.268, mae_surf_Uy=1.030, mae_surf_p=15.442
- val_tandem_transfer: loss=0.8994, mae_surf_p=29.2
- val_ood_cond: loss=0.2262, mae_surf_p=8.9
- val_ood_re: loss=0.2233, mae_surf_p=7.6

**GALE diagnostic — ||gale_out_proj.weight||_F per block:**
- Block 0: 15.518
- Block 1: 16.370
- Block 2: 21.689

**Seed 73: Not run** — seed 42 showed clear regression across all metrics, no point in second seed.

**W&B:** run ID `8b6u00qn`
**Peak VRAM:** 49.8 GB (vs ~45.9 GB baseline, +3.9 GB from GALE cross-attention)
**Command:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/geotransolver-gale-s42" \
  --wandb_group "geotransolver-gale" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --geotransolver_gale --gale_latent_dim 32 \
  --seed 42
```

## Analysis

**Clear negative result.** GALE geometry-latent cross-attention degraded surface pressure MAE by +23.6% on in-distribution and +16.8% on OOD conditions. The p_tan target showed minimal change (+2.4%), so the geometry conditioning did not help with the NACA6416 OOD shape generalization at all.

**Key observations:**
1. **GALE out_proj weights grew large** (Frobenius norms 15-22), meaning the cross-attention learned to contribute significantly — but in a harmful direction. The zero-initialization didn't prevent learning harmful features over 126 epochs; it just delayed them.
2. **VRAM overhead was significant** (+3.9 GB), making this an expensive addition even if it had worked.
3. **Hypothesis failure mode:** The geometry latent pools raw input features over surface nodes, but these features are already heavily processed by the existing pipeline (saf, dsdf, TE coords, curvature). The cross-attention likely introduced a redundant but noisy signal path that competed with the existing slice-attention routing mechanism.
4. **The val_loss trajectory** (0.412 at best vs baseline ~0.39) shows GALE consistently added ~5% overhead to overall loss, not just surface pressure.

**Why it didn't work:** The Transolver's slice-attention already captures geometry implicitly through the spatial bias and dsdf features. Injecting a separate geometry latent via cross-attention creates a competing information pathway that the model has to reconcile. On a small 3-block architecture with only 126 training epochs, this reconciliation hurts rather than helps.

## Suggested follow-ups
- **Geometry latent as a conditioning signal to AdaLN** instead of cross-attention — lighter touch, modulates existing features rather than competing with them
- **Frozen geometry encoder** (pretrained on a shape classification task) so the latent is meaningful from epoch 0 rather than learned end-to-end
- **Smaller latent dim** (e.g., 8 instead of 32) to force a more compressed shape representation
- **GALE only on the last block** rather than all blocks, to limit interference with learned slice routing

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2216 because ADVISOR: Closing as dead end. All four surface MAE metrics regressed significantly — p_in +23.6%, p_oodc +16.3%, p_re +19.1%, p_tan +2.4%. The GALE cross-attention mechanism creates a competing information pathway that hurts the Transolver's existing slice-attention. Your analysis is excellent — the architecture already implicitly captures geometry via spatial bias and dsdf features, making the explicit geometry latent redundant and noisy.

Your suggested follow-ups (AdaLN conditioning, frozen geometry encoder, smaller latent dim, last-block-only) are thoughtful, but given the magnitude of regression across ALL metrics, I think more promising opportunities lie in other directions. I'll assign you a fresh hypothesis shortly..
